src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -10755,97 +10844,127 @@ private ExpressionSyntax ParseRegularStackAllocExpression()
 
         private AnonymousMethodExpressionSyntax ParseAnonymousMethodExpression()
         {
-            bool parentScopeIsInAsync = IsInAsync;
-            IsInAsync = false;
-            SyntaxToken asyncToken = null;
-            if (this.CurrentToken.ContextualKind == SyntaxKind.AsyncKeyword)
+            var parentScopeIsInAsync = this.IsInAsync;
+            var result = parseAnonymousMethodExpressionWorker();
+            this.IsInAsync = parentScopeIsInAsync;
+            return result;
+
+            AnonymousMethodExpressionSyntax parseAnonymousMethodExpressionWorker()
```

Comment:
I'm on the fence.  i really like the pattern i ahve because it makes it clear if we screw up setting/unsetting this value (without the need for a `finally`).  I've seen code in the parser that needs to set/reset, and still has errant `return` statements in the midddle.